app/plugin.go

```diff
@@ -182,6 +187,74 @@ func (a *App) InitPlugins(pluginDir, webappPluginDir string) {
 	a.SyncPluginsActiveState()
 }
 
+// SyncPlugins synchronizes the plugins installed locally
+// with the plugin bundles available in the file store.
+func (a *App) SyncPlugins() *model.AppError {
+	mlog.Info("Synching plugins with the file store")
+
+	pluginDirs, err := ioutil.ReadDir(*a.Config().PluginSettings.Directory)
+	if err != nil {
+		return model.NewAppError("SyncPlugins", "app.plugin.sync.read_local_folder.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Remove plugins locally to sync with file store.
+	for _, dir := range pluginDirs {
+		if !dir.IsDir() {
+			continue
+		}
+
+		// Only handle managed plugins with .filestore flag file.
+		_, err := os.Stat(filepath.Join(*a.Config().PluginSettings.Directory, dir.Name(), managedPluginFileName))
+
+		if os.IsNotExist(err) {
+			mlog.Warn("Found unmanaged plugin. Ignoring in plugin sync with the filestore.", mlog.String("plugin", dir.Name()))
+
+		} else if err != nil {
+			mlog.Error("Error reading local plugin directoy. Skipped for sync.", mlog.String("folder", dir.Name()), mlog.Err(err))
+
+		} else {
+			mlog.Debug("Plugin Sync: Uninstalling plugin locally", mlog.String("plugin", dir.Name()))
+			if err := a.removePluginLocally(dir.Name()); err != nil {
+				mlog.Error("Plugin Sync: Error uninstalling managed plugin.", mlog.String("plugin", dir.Name()), mlog.Err(err))
+			}
+		}
+	}
+
+	// Install plugins from the file store.
+	exists, fileErr := a.FileExists(fileStorePluginFolder)
+	if fileErr != nil {
+		return model.NewAppError("SyncPlugins", "app.plugin.sync.check_filestore.app_error", nil, fileErr.Error(), http.StatusInternalServerError)
+	}
+
+	if !exists {
+		return nil
+	}
+
+	fileStorePaths, listDirErr := a.ListDirectory(fileStorePluginFolder)
+	if listDirErr != nil {
+		return model.NewAppError("SyncPlugins", "app.plugin.sync.list_filestore.app_error", nil, listDirErr.Error(), http.StatusInternalServerError)
+	}
+
+	for _, path := range fileStorePaths {
+		if strings.HasSuffix(path, ".tar.gz") {
+			reader, fileReaderErr := a.FileReader(filepath.Join("./", path))
+			defer reader.Close()
```

Comment:
this will probably cause a panic if `failReaderErr` is not nil. `defer reader.Close()` should go after the `fileReaderErr != nil` check

Reply:
Good catch. Fixed